### PR TITLE
Relax some libcontainerd client locking

### PR DESCRIPTION
**Pull request by cpuguy83**
Fixes https://github.com/moby/moby/issues/36798

Release lock on `Restore` while interacting with containerd.
Also adds some timeouts on the contexts on startup.

### CODE REVIEW

File: libcontainerd/client_daemon.go [hunk 1]
```diff
@@ -131,9 +131,28 @@ func (c *client) Version(ctx context.Context) (containerd.Version, error) {
 	return c.getRemote().Version(ctx)
 }
 
+// Restore restores loads the containerd container.
```
Comment (mlaventure) [inline libcontainerd/client_daemon.go]: `s/restores//`

**Review by mlaventure — APPROVED** [review]
one nit, LGTM.

It should allow other threads to proceed. Although I don't understand how/why the one you highlighted in the issue never gets an answer from containerd (which would unlock it).

Hum, may be because of the fifo being already closed when containerd tries to open then, in which case the context timeout fixes that too :thinking:

**Comment by cpuguy83** [comment]
Yes the timeout definitely fixes it, though it blocks everything else until the timeout is reached.

**Review by vdemeester — APPROVED** [review]
LGTM :lion:

**Comment by tonistiigi** [comment]
What are these arbitrary 60 sec timeouts for? We shouldn't hide hangs like this and leave daemon in an undefined state.

If there is some fifo hanging it should be fixed instead. We have tools for force closing fifos one-sided. `0.2` branch used `discardFifo` helper that I don't see any more.

**Comment by cpuguy83** [comment]
@tonistiigi I think a hung daemon is an undefined state. Where as an error saying the deadline exceeded can provide much more valuable information because we get the call stack in the error rather than requiring users to fetch the stack.

**Comment by mlaventure** [comment]
I think the issue on containerd side should have been fixed by https://github.com/containerd/containerd/pull/2229

**Comment by mlaventure** [comment]
Hum, but that went into 1.0.x so I think it was probably there when @stevvooe got his issue.

**Comment by cpuguy83** [comment]
I don't see it as temporary. It protects the daemon from a range of potential bugs.
A stuck daemon, especially stuck in startup is really bad.

**Comment by tonistiigi** [comment]
The problem here isn't so much the timeout but ignoring this range of potential bugs. Regarding containerd bugs this isn't really different than most other places where we call to containerd and expect it to answer. I have nothing against detecting unexpected behavior from containerd and handling it better in moby (for example marking containers as "errored/unresponsive/corrupted", refusing to handle specific API calls etc.

**Comment by cpuguy83** [comment]
@tonistiigi I will remove the timeout for now, however I'm not sure I agree with the assessment that it's hiding bugs. It's exposing bugs in a much easier to debug way that doesn't block production systems.
But it would be nice indeed if we still had some sort of status on the container rather than just hiding the affected container.

**Comment by thaJeztah** [comment]
Failure on PowerPC; could be a flake; https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/9465/console

```
16:48:06 === RUN   TestNetworkLoopbackNat
16:48:07 --- FAIL: TestNetworkLoopbackNat (0.94s)
16:48:07 	nat_test.go:86: assertion failed: it works (msg string) !=  (string)
```

**Comment by tonistiigi** [comment]
LGTM

**Comment by thaJeztah** [comment]
Looks like we have enough LGTM's - merging

**Pull request merged**